torch/csrc/jit/script/compiler.cpp

```diff
@@ -1070,44 +1070,64 @@ struct to_ir {
     return emitIfExpr(expr.range(), cond_value, true_expr, false_expr);
   }
 
-  // emit a single expr from the loop comprehension so that we can correctly
-  // type the list we create, then remove the nodes we emitted
-  TypePtr getListCompType(
-      const ListComp& lc,
-      const ListTypePtr& input_list_type) {
+  TypePtr emitInTempEnvironment(const std::function<TypePtr()>& fn) {
     auto b = graph->insertNode(graph->create(prim::Loop))->addBlock();
     pushFrame(b);
     WithInsertPoint guard(b);
-    auto li_elem = graph->insertNode(
-        graph->createUninitialized(input_list_type->getElementType()));
-    emitExprsAssign(
-        List<Expr>::create(lc.range(), {lc.target()}),
-        {std::make_shared<SimpleValue>(li_elem->output())},
-        lc.range(),
-        /*n_binders*/ 1);
-    auto ret_type = emitExpr(lc.elt())->type();
+    auto ret_type = fn();
     popFrame();
     b->owningNode()->destroy();
     return ret_type;
   }
 
+  // emit a single expr from the loop comprehension so that we can correctly
+  // type the list we create, then remove the nodes we emitted
+  TypePtr getListCompType(const ListComp& lc, const TypePtr type_ptr) {
+    return emitInTempEnvironment([&]() {
+      auto li_elem = graph->insertNode(graph->createUninitialized(type_ptr));
+      emitExprsAssign(
+          List<Expr>::create(lc.range(), {lc.target()}),
+          {std::make_shared<SimpleValue>(li_elem->output())},
+          lc.range(),
+          /*n_binders*/ 1);
+      return emitExpr(lc.elt())->type();
+    });
+  }
+
+  // emit a single index of the Iterable to get the children type
+  TypePtr getIterableChildrenType(const ListComp& lc, IterableTree& tree) {
+    return emitInTempEnvironment([&]() {
+      return tree.getitem(lc.range(), method, graph->insertConstant(0))->type();
```

Comment:
What happens if there's no elements in the `IterableTree`?

Reply:
`IterableTree` either consists of list, string, or range. To index into 0 would be a runtime error not a compile time error, but the index never gets evaluated at runtime because the nodes are removed. I can add a test